pkg/acpi/raw.go

```diff
@@ -0,0 +1,139 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package acpi
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/u-root/u-root/pkg/memio"
+)
+
+// Raw ACPI table support. Raw tables are those tables
+// one needs to read in, write out, but not change in any way.
+// This is needed when, e.g., there is need to create files for
+// coreboot cbfs.
+
+// Raw is just a table embedded in a []byte.  Operations on Raw are
+// for figuring out how to skip a table you don't care about or, possibly,
+// truncating a table and regenerating a checksum.
+type Raw struct {
+	data []byte
+}
+
+var _ = Table(&Raw{})
+
+// NewRaw returns a new Raw table given a byte slice.
+func NewRaw(b []byte) ([]Table, error) {
+	var tab []Table
+	for {
+		if len(b) == 0 {
+			break
+		}
+		if len(b) < headerLength {
+			return nil, fmt.Errorf("NewRaw: byte slice is only %d bytes and must be at least %d bytes", len(b), headerLength)
+		}
+		u := binary.LittleEndian.Uint32(b[lengthOffset : lengthOffset+4])
+		if int(u) > len(b) {
+			return nil, fmt.Errorf("Table length %d is too large for %d byte table (Signature %q", u, len(b), string(b[0:4]))
+		}
+		tab = append(tab, &Raw{data: b[:u]})
+		b = b[u:]
+	}
+	return tab, nil
+}
+
+func RawFromFile(r io.Reader) ([]Table, error) {
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	return NewRaw(b)
+}
+
+// RawFromFile reads a raw table in from a file.
+func RawFromName(n string) ([]Table, error) {
+	f, err := os.Open(n)
+	if err != nil {
+		return nil, err
+	}
+	return RawFromFile(f)
+}
+
+// ReadRaw reads a full table in, given an address.
+// ReadRaw uses the io package. This may not always work
+// if the kernel has restrictions on reading memory above
+// the 1M boundary, and the tables are above boundary.
+func ReadRaw(a int64) (Table, error) {
+	var u memio.Uint32
+	// Read the table size at a+4
+	if err := memio.Read(a+4, &u); err != nil {
+		return nil, err
+	}
+	Debug("ReadRaw: Size is %d", u)
+	dat := memio.ByteSlice(make([]byte, u))
+	if err := memio.Read(a, &dat); err != nil {
+		return nil, err
+	}
+	return &Raw{data: []byte(dat)}, nil
+}
+
+// Data returns all the data in a Raw table.
+func (r *Raw) Data() []byte {
+	return r.data
+}
+
+// TableData returns the Raw table, minus the standard ACPI header.
+func (r *Raw) TableData() []byte {
+	return r.data[headerLength:]
+}
+
+// Sig returns the table signature.
+func (r *Raw) Sig() string {
+	return fmt.Sprintf("%s", r.data[:4])
+}
+
+// Len returns the total table length.
+func (r *Raw) Len() uint32 {
+	return uint32(len(r.data))
+}
+
+// Revision returns the table Revision.
+func (r *Raw) Revision() uint8 {
+	return uint8(r.data[8])
+}
+
+// CheckSum returns the table CheckSum.
+func (r *Raw) CheckSum() uint8 {
+	return uint8(r.data[9])
+}
+
+// OEMID returns the table OEMID.
+func (r *Raw) OEMID() string {
+	return fmt.Sprintf("%q", r.data[10:16])
```

Comment:
Why %q? Should just be %s, here and below?

Reply:
I've found arbitrary garbage in these things -- spaces, control chars -- so it's best to use %q.